### PR TITLE
rgw: mfa support

### PR DESCRIPTION
**Pull request by yehudasa**
- otp object class implementation
 - rgw support for S3 mfa api
 - rgw otp metadata provider to allow sync of otp data across zones
 - sync token due to time skew
 - support hex and base32 seeds

### CODE REVIEW

File: src/cls/otp/cls_otp.cc [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+  result->result = OTP_CHECK_UNKNOWN;
+  result->timestamp = now;
+}
+
+static int get_otp_instance(cls_method_context_t hctx, const string& id, otp_instance *instance)
+{
+  bufferlist bl;
+  string key = otp_key_prefix + id;
+ 
+  int r = cls_cxx_map_get_val(hctx, key.c_str(), &bl);
```
Comment (cbodley) [inline src/cls/otp/cls_otp.cc]: the `cls_cxx_` apis all take `std::string` - so these are converting to `const char*` and back again

File: src/cls/otp/cls_otp_ops.h [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+#include "include/types.h"
+#include "include/utime.h"
+#include "cls/otp/cls_otp_types.h"
+
+struct cls_otp_set_otp_op
+{
+  list<rados::cls::otp::otp_info_t> entries;
+
+  cls_otp_set_otp_op() = default;
```
Comment (cbodley) [inline src/cls/otp/cls_otp_ops.h]: the compiler will generate this for you "if no other constructor is explicitly declared" - https://en.wikipedia.org/wiki/Special_member_functions

File: src/cls/otp/cls_otp.cc [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+  bool verify(const ceph::real_time& timestamp, const string& val);
+
+  void find(const string& token, otp_check_t *result);
+};
+WRITE_CLASS_ENCODER(otp_instance)
+
+
+void otp_instance::trim_expired(const ceph::real_time& now)
+{
+  ceph::real_time window_start = now - make_timespan(otp.step_size);
```
Comment (cbodley) [inline src/cls/otp/cls_otp.cc]: `make_timespan()` takes a double, so this does an extra conversion. `std::chrono::seconds(otp.step_size)` may be more concise here?

File: src/cls/otp/cls_otp_ops.h [hunk 4]
```diff
@@ -0,0 +1,150 @@
+#ifndef CEPH_CLS_OTP_OPS_H
+#define CEPH_CLS_OTP_OPS_H
+
+#include "include/types.h"
+#include "include/utime.h"
+#include "cls/otp/cls_otp_types.h"
+
+struct cls_otp_set_otp_op
+{
+  list<rados::cls::otp::otp_info_t> entries;
```
[the remark this answers is not in the document]
Comment (yehudasa) [reply src/cls/otp/cls_otp_ops.h]: @cbodley only iterating over it, never access entries via index. list.size() should be O(1) in c++11 (and it is in all relevant implementations).

[the remark this answers is not in the document]
Comment (cbodley) [reply src/cls/otp/cls_otp_ops.h]: i worry about the complexity of allocations though, where list is linear and vector is logarithmic

[the remark this answers is not in the document]
Comment (yehudasa) [reply src/cls/otp/cls_otp_ops.h]: @cbodley it's pretty moot as the number of entries expected to be here are somewhere between zero and one, but I can change it to vector if you really insist.

**Comment by cbodley** [comment]
looking good 👍 do you have plans to add some test coverage?

**Comment by yehudasa** [comment]
@cbodley yes, although it's not clear to me which tool is the best one. Requires functional tests intermixed with admin stuff.

File: CMakeLists.txt [hunk 5]
```diff
@@ -368,6 +368,8 @@ if(WITH_RADOSGW)
   endif()
 endif(WITH_RADOSGW)
 
+find_package(liboath REQUIRED)
```
Comment (cbodley) [inline CMakeLists.txt]: i'm guessing that we'll also need to express this dependency in `install-deps.sh` and the rpm/deb builds as well?

**Comment by yehudasa** [comment]
@cbodley can this be merged?
http://pulpito.ceph.com/yehudasa-2018-03-23_17:18:06-rgw-wip-rgw-mfa-distro-basic-smithi/

File: src/rgw/rgw_admin.cc [hunk 6]
```diff
@@ -2373,6 +2396,52 @@ int create_new_bucket_instance(RGWRados *store,
...
+                             pins[0].c_str());
+    if (rc != OATH_INVALID_OTP) {
+      rc = oath_totp_validate2(totp.seed_bin.c_str(), totp.seed_bin.length(),
+                               start_time, 
+                               step_size,
+                               time_ofs - step_size, /* smaller time_ofs moves time forward */
+                               1,
+                               nullptr,
+                               pins[1].c_str());
+      if (rc != OATH_INVALID_OTP) {
```
Comment (cbodley) [inline src/rgw/rgw_admin.cc]: from the oath reference manual on [oath_totp_validate2](https://www.nongnu.org/oath-toolkit/liboath-api/liboath-oath.html#oath-totp-validate2):
> Returns absolute value of position in OTP window (zero is first position), or OATH_INVALID_OTP if no OTP was found in OTP window, or an error code.

shouldn't this check for other errors codes as well?

and is there something we can do with the position it returns on success to calculate a more accurate offset?

File: src/rgw/rgw_admin.cc [hunk 7]
```diff
@@ -7259,5 +7351,239 @@ int main(int argc, const char **argv)
...
+    }
+
+    if (totp_serial.empty()) {
+      cerr << "ERROR: TOTP device serial number was not provided (via --totp-serial)" << std::endl;
+      return EINVAL;
+    }
+
+    if (totp_pin.size() != 2) {
+      cerr << "ERROR: missing two --totp-pin params (--totp-pin=<first> --totp-pin=<second>)" << std::endl;
+    }
```
Comment (cbodley) [inline src/rgw/rgw_admin.cc]: missing a `return EINVAL`? this would hit an assert in `scan_totp()`

File: src/rgw/rgw_admin.cc [hunk 8]
```diff
@@ -6203,7 +6295,7 @@ int main(int argc, const char **argv)
       cerr << "ERROR: failed to read input: " << cpp_strerror(-ret) << std::endl;
       return -ret;
     }
-    ret = store->meta_mgr->put(metadata_key, bl, RGWMetadataHandler::APPLY_ALWAYS);
+    ret = store->meta_mgr->put(metadata_key, bl, RGWMetadataHandler::RGWMetadataHandler::APPLY_ALWAYS);
```
Comment (cbodley) [inline src/rgw/rgw_admin.cc]: ^ this doesn't look right

**Comment by cbodley** [comment]
@yehudasa can you please add some user docs to show how to set this up?

**Pull request merged**